crates/shared/src/price_estimation/trade_verifier.rs

```diff
@@ -553,68 +733,203 @@ enum Error {
+            let (in_amount, out_amount) = match kind {
+                OrderKind::Sell => (500u64.into(), 250u64.into()),
+                OrderKind::Buy => (250u64.into(), 500u64.into()),
+            };
+            let price_query = PriceQuery {
+                in_amount: NonZeroU256::new(in_amount).unwrap(),
+                kind,
+                sell_token,
+                buy_token,
+            };
+
+            (price_query, out_amount)
+        };
+
+        // The jit order should fully fill the query
+        let jit_executed_amount = |side: &dto::Side| -> U256 {
+            match side {
+                dto::Side::Sell => 250u64.into(),
+                dto::Side::Buy => 500u64.into(),
+            }
+        };
+
+        let jit_order = |side: dto::Side| -> dto::JitOrder {
+            dto::JitOrder {
+                sell_token: buy_token,                       // Solver sells buy token
+                buy_token: sell_token,                       // Solver buys sell token
+                executed_amount: jit_executed_amount(&side), // Fully fills the query
+                side,
+                sell_amount: 250u64.into(),
+                buy_amount: 500u64.into(),
+                receiver: H160::zero(),
+                valid_to: 0,
+                app_data: AppDataHash::default(),
+                partially_fillable: false,
+                sell_token_source: SellTokenSource::Erc20,
+                buy_token_destination: BuyTokenDestination::Erc20,
+                signature: vec![],
+                signing_scheme: SigningScheme::Eip1271,
+            }
+        };
+
+        let test_cases = [
+            (
+                query_with_out_amount(OrderKind::Sell),
+                jit_order(dto::Side::Sell),
+            ),
+            (
+                query_with_out_amount(OrderKind::Buy),
+                jit_order(dto::Side::Buy),
+            ),
+            (
+                query_with_out_amount(OrderKind::Sell),
+                jit_order(dto::Side::Buy),
+            ),
+            (
+                query_with_out_amount(OrderKind::Buy),
+                jit_order(dto::Side::Sell),
+            ),
+        ];
+
+        for ((query, out_amount), jit_order) in test_cases {
+            let trade_kind = TradeKind::Regular(Trade {
+                clearing_prices: clearing_prices.clone(),
+                gas_estimate: Some(50_000),
+                pre_interactions: vec![],
+                interactions: vec![],
+                solver: H160::from_low_u64_be(0x1234),
+                tx_origin: None,
+                jit_orders: vec![jit_order.clone()],
+            });
+
+            // The Settlement contract is not expected to lose any tokens, but we lose a bit
+            // to test the inaccuracy limit.
+            let tokens_lost = hashmap! {
+                sell_token => BigRational::from_integer(50.into()),
+                buy_token => BigRational::from_integer(25.into()),
```

Comment:
nit: Would be slightly nicer to test the inaccuracy threshold of sell and buy token individually.
The current test wouldn't show a bug where we forget to test the accuracy of one of the tokens by accident because it's set up that both tokens pass or fail the threshold at once.
But I'm fine with delaying this to merge this PR sooner.